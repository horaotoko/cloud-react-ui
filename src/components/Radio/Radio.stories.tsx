import { Story, Meta } from '@storybook/react';
import { Radio, Props } from './Radio';

export default {
  component: Radio,
  title: 'components/Radio',
} as Meta;

const Template: Story<Props> = (args) => <Radio {...args}>Radio</Radio>;

export const Default = Template.bind({
  id: 'radio-1',
  name: 'default',
  value: 'value-1',
});

export const Disabled = Template.bind({});
Disabled.args = { disabled: true, id: 'radio-2', name: 'disabled' };
