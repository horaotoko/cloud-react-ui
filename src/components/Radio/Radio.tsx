import { forwardRef, ComponentPropsWithRef } from 'react';
import styled, { css } from 'styled-components';
import { defaultProps } from '../../theme';

export type Props = {
  id: string;
} & ComponentPropsWithRef<'input'>;

const RadioWrapper = styled.div`
  display: inline-flex;
  align-items: center;
`;
RadioWrapper.defaultProps = defaultProps;

const Input = styled.input`
  /* visually-hidden */
  /* TODO: Should extract visually-hidden style as a utility style */
  clip: rect(0 0 0 0);
  clip-path: inset(50%);
  height: 1px;
  overflow: hidden;
  position: absolute;
  white-space: nowrap;
  width: 1px;
`;
Input.defaultProps = defaultProps;

const RadioButton = styled.span`
  display: grid;
  place-items: center;
  margin-right: 4px;

  ${({ theme: { radio } }) => css`
    --radio-state-color: ${radio.backgroundColor.unchecked};

    width: 16px;
    height: 16px;
    background-color: #fff;
    border-radius: ${radio.borderRadius};
    border: 5px solid var(--radio-state-color);

    ${Input}:checked + ${Label} > & {
      --radio-state-color: ${radio.backgroundColor.checked};
    }

    ${Input}:disabled + ${Label} > & {
      --radio-state-color: ${radio.backgroundColor.disabled};
    }
  `}
`;
RadioButton.defaultProps = defaultProps;

const Label = styled.label`
  cursor: pointer;
  display: inline-flex;
  align-items: center;

  ${Input}:disabled + & {
    cursor: not-allowed;
    color: #999;
  }
`;
const LabelText = styled.span``;

const Radio = forwardRef<HTMLInputElement, Props>(
  (
    {
      children,
      className,
      id,
      name,
      value,
      disabled,
      defaultChecked,
      onClick,
      onChange,
      ...rest
    },
    ref
  ) => (
    <RadioWrapper>
      <Input
        type="radio"
        id={id}
        ref={ref}
        name={name}
        value={value}
        disabled={disabled}
        defaultChecked={defaultChecked}
        onClick={onClick}
        onChange={onChange}
        {...rest}
      />
      <Label htmlFor={id}>
        <RadioButton />
        <LabelText>{children}</LabelText>
      </Label>
    </RadioWrapper>
  )
);
Radio.displayName = 'Radio';

export { Radio };
