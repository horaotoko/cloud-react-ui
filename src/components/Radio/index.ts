export { Radio } from './Radio';
export type { Props as RadioProps } from './Radio';
