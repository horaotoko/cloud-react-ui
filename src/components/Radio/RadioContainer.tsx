import { RadioPresenter } from "./RadioPresenter";

export type RadioProps = {
  name?: string;
  checked?: boolean;
  value?: string;
  disabled?: boolean;
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onClick?: (event: React.MouseEvent<HTMLInputElement, MouseEvent>) => void;
  children?: React.ReactNode;
  className?: string;
};

export function RadioContainer(props: RadioProps): React.ReactElement {
  return <RadioPresenter {...props} />;
}
