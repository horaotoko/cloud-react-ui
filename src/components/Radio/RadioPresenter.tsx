import styled, { css } from "styled-components";
import { defaultProps } from "../../theme";
import { RadioProps } from "./RadioContainer";

const RadioWrapper = styled.span`
  ${({ theme: { radio } }) => css`
    position: relative;
    display: inline-block;
    width: ${radio.width};
    height: ${radio.height};
    vertical-align: middle;
  `}
`;
RadioWrapper.defaultProps = defaultProps;

const Input = styled.input`
  ${({ theme: { radio } }) => css`
    opacity: 0;
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    margin: 0;
    cursor: pointer;

    :checked + ${Box} {
      background-color: ${radio.backgroundColor.checked};
      border-width: ${radio.borderWidth.checked};
      border-style: ${radio.borderStyle.checked};
      border-color: ${radio.borderColor.checked};
    }

    :disabled + ${Box} {
      background-color: ${radio.backgroundColor.disabled};
      border-width: ${radio.borderWidth.disabled};
      border-style: ${radio.borderStyle.disabled};
      border-color: ${radio.borderColor.disabled};
    }
  `}
`;
Input.defaultProps = defaultProps;

const Box = styled.span`
  ${({ theme: { radio } }) => css`
    position: absolute;
    width: 100%;
    height: 100%;
    border-radius: ${radio.borderRadius};
    background-color: ${radio.backgroundColor.unchecked};
    border-width: ${radio.borderWidth.unchecked};
    border-style: ${radio.borderStyle.unchecked};
    border-color: ${radio.borderColor.unchecked};
    pointer-events: none;
  `}
`;
Box.defaultProps = defaultProps;

const IconWrapper = styled.span`
  ${({ theme: { radio } }) => css`
    position: absolute;
    display: inline-block;
    top: 50%;
    left: 50%;
    width: ${radio.icon.width};
    height: ${radio.icon.height};
    transform: translate(-50%, -50%);
    pointer-events: none;

    & > svg {
      vertical-align: top;
    }
  `}
`;
IconWrapper.defaultProps = defaultProps;

const Icon = styled(({ className }) => (
  <svg
    xmlns="http://www.w3.org/2000/svg"
    viewBox="0 0 6 6"
    className={className}
  >
    <circle
      cx="102"
      cy="546"
      r="3"
      fillRule="evenodd"
      transform="translate(-99 -543)"
    />
  </svg>
))`
  ${({ theme: { radio } }) => css`
    width: ${radio.icon.width};
    height: ${radio.icon.height};

    & > circle {
      fill: ${radio.icon.color};
    }
  `}
`;
Icon.defaultProps = defaultProps;

const Label = styled.label`
  cursor: pointer;
  vertical-align: middle;
`;

export const RadioPresenter: React.FC<RadioProps> = ({
  children,
  className,
  ...rest
}) => {
  return (
    <Label>
      <RadioWrapper className={className}>
        <Input type="radio" {...rest} />
        <Box />
        <IconWrapper>
          <Icon />
        </IconWrapper>
      </RadioWrapper>

      {children}
    </Label>
  );
};
