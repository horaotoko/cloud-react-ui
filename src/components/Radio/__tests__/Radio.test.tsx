import { render, fireEvent } from "@testing-library/react";
import { Radio } from "../Radio";

describe("Radio", () => {
  it("default", () => {
    const { asFragment, getByRole } = render(<Radio>radio</Radio>);
    expect(asFragment()).toMatchSnapshot();

    const radio = getByRole("radio");
    expect(radio).not.toBeChecked();
    expect(radio).not.toBeDisabled();
  });

  it("defaultChecked props", () => {
    const { getByRole } = render(<Radio defaultChecked={true}>radio</Radio>);
    expect(getByRole("radio")).toBeChecked();
  });

  it("disabled props", () => {
    const { getByRole } = render(<Radio disabled={true}>radio</Radio>);
    expect(getByRole("radio")).toBeDisabled();
  });

  it("onChange, onClick props", () => {
    const onChange = jest.fn();
    const onClick = jest.fn();
    const { getByText } = render(
      <Radio onChange={onChange} onClick={onClick}>
        radio
      </Radio>
    );
    fireEvent.click(getByText("radio"));

    expect(onChange).toBeCalledTimes(1);
    expect(onClick).toBeCalledTimes(1);
  });
});
