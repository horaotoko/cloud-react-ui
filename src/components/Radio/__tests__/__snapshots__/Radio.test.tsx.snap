// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio default 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  height: 1px;
  overflow: hidden;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.c6 {
  display: grid;
  place-items: center;
  margin-right: 4px;
  --radio-state-color: #FFFFFF;
  width: 16px;
  height: 16px;
  background-color: #fff;
  border-radius: 8px;
  border: 5px solid var(--radio-state-color);
}

.c1:checked + .c3 > .c5 {
  --radio-state-color: #3B7DE9;
}

.c1:disabled + .c3 > .c5 {
  --radio-state-color: #D4D8DD;
}

.c4 {
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1:disabled + .c3 {
  cursor: not-allowed;
  color: #999;
}

<div
    class="c0"
  >
    <input
      class="c1 c2"
      id="radio-1"
      type="radio"
      value=""
    />
    <label
      class="c3 c4"
      for="radio-1"
    >
      <span
        class="c5 c6"
      />
      <span
        class=""
      >
        radio
      </span>
    </label>
  </div>
</DocumentFragment>
`;
